### ISSUE TITLE
Shuffled images in dataloader using custom data

### ISSUE THREAD

**Issue by AidanAbramson** [ultralytics/ultralytics#442]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training, Detection

### Bug

Hi!
I am trying to train the yolov8n model on custom data. Using the Python API, I have provided a .yaml file containg paths to .txt files which each contain a list of paths to the images.

The dataset is read correctly and the labels are cached and read as it should but when it creates the batch mosaic the bounding boxes are completely wrong. 

After following the code for ~2hrs I have concluded that the problem comes from the image loading (after disabling augmentations and such to find the source of the problem). 
For some reason, the image at index 0 (batch['imgs'][0]) does not match with the image path at index 0 (batch['img_file'][0]) and therefore with the bounding box at index 0 (batch['bboxes'][0]). To clarify, the bounding box does match the file path, but not the image itself.

Could there be any kind of shuffling happening only to the images after loading? I followedthe code in the trainer.py script and it seems the problem happens in the dataloader itself, but I could not manage to follow the code into the dataloader so I couldn't find if the problem is from there...

Any help is appreciated! Thanks.

### Environment

torch 1.13, Nvidia 3080Ti, cuda v11.4
python 3.7.15
Ubuntu v18.04

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Robotatron** [comment]
do you have image rotation augmentation enabled? Could be that

**Comment by AidanAbramson** [comment]
@Robotatron no, I have disabled all augmentations just to check what's the problem. But like I said, the bounding box just doesn't match the first image, but it dies match the first image file path, which when I display, is not the same as the first image.

**Comment by AidanAbramson** [comment]
@Robotatron no, I have disabled all augmentations just to check what's the problem. But like I said, the bounding box just doesn't match the first image, but it dies match the first image file path, which when I display, is not the same as the first image.

**Comment by Robotatron** [comment]
So do the bboxes break only with mosaic? Do bboxes are drawn correctly without mosaic augmentation? Otherwise check your data in FiftyOne to see if bboxes are correct in the annotations

**Comment by AidanAbramson** [comment]
I am not familiar with FiftyOne so I'll have to give that a look. its not just the mosaic, the batch itself seemed to have only shuffled the img data (the numpy arrays) but not the bboxs, making some of them not match...

**Comment by glenn-jocher** [comment]
@AidanAbramson 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://github.com/ultralytics/yolov5/issues/36) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths) and 2) a class `names` dictionary:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes (80 COCO classes)
names:
  0: person
  1: bicycle
  2: car
  ...
  77: teddy bear
  78: hair drier
  79: toothbrush

```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<p align="center"><img width="750" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg"></p>

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>

Good luck 🍀 and let us know if you have any other questions!

**Comment by AidanAbramson** [comment]
HI, thanks for the reply!
My data is scoured throughout my server and i have a .txt file which contains the paths to every file, I used yolov7 before I used this and I had to edit the dataset.py script in their repository to accept this kind of format. When i tried to do this here it seemed to have worked. 
Other than that evryhting is in the yolov5 format. Could me using .txt files instead of directories be the reason for the matching not working? If so could you maybe direct me to the code I can edit to do so? Or is it best I just copy all files to directories and use the default format?
Thanks again.

**Comment by AidanAbramson** [comment]
Hi, @glenn-jocher. Would appreciate you replying back. Thanks.

**Comment by AidanAbramson** [comment]
After downloading the new update and removing shuffle = mode == "train" in line 59 in builder.py and adding shuffle=False, the problem seems to be fixed, so obviously a problem with shuffling. Leaving this open in case any of you want to take a look.

**Comment by AidanAbramson** [comment]
update: when shuffle is true, which it is by default when it's in training mode, and the data provided are .txt files, the shuffling of the images do not match the bounding boxes.

This is probably a bug...

Steps to reproduce: create .txt files containing the paths to your images. After creating a matching .yaml file train the model without pretrained weights through the python API. when the first epoch starts, check the train grid created in the training directory. Thanks.

**Comment by pderrenger** [comment]
Hi @AidanAbramson, thanks for your valuable feedback. It seems related to the shuffling behavior when using .txt files as input. I have written this down as a potential bug for further investigation. The use of .txt files as data sources should not affect the correct matching of bounding boxes, and shuffling should not lead to a mismatch between images and their corresponding bounding boxes.

As a temporary workaround, removing shuffling by setting shuffle=False appears to resolve the issue. I have escalated this to our development team for further review. We will keep you updated on the progress.

Your patience and cooperation are deeply appreciated. If you encounter any other roadblocks or have additional details to share, please feel free to do so. Thank you!
